### PR TITLE
feat: add loom test for the counter

### PR DESCRIPTION
**Pull request by BowenXiao1999**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Use https://github.com/tokio-rs/loom for concurrency test. Maybe can make the correctness more confident. But I'm not so sure how to write the best test.

it requires some code change for previous structure, so may need discuss:
* remove workspacke hack in task_stats_alloc crate. Otherwise there will be package conflict if you run `RUSTFLAGS="--cfg loom" cargo test --test loom`. Have not investigate deeply but simple remove just works.

* Refactor `&'static AtomicUsize` to be `NonNull<AtomicUsize>`. This is because the AtomicUsize in loom type do not support `.as_mut_ptr` (described in https://github.com/tokio-rs/loom/issues/298), so we use NonNull as intermediate workaround, that will add some unsafe code in `.add()` and `.sub()`. But seems OK.

* To test the drop, I have to add a flag var AtomicUsize in `.sub()` to ensure that the value is dropped. Please provide some suggestions if you have better ideas on how to write test.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/utils/task_stats_alloc/Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,10 @@ tokio = { version = "0.2", package = "madsim-tokio", features = [
     "time",
     "signal",
 ] }
-workspace-hack = { path = "../../workspace-hack" }
+#workspace-hack = { path = "../../workspace-hack" }
```
Comment (wangrunji0408) [inline src/utils/task_stats_alloc/Cargo.toml:18]: We can either disable it on loom:

```suggestion
[target.'cfg(not(loom))'.dependencies]
workspace-hack = { path = "../../workspace-hack" }
```

or just remove it:

```suggestion
```

Personally I prefer to remove it.

File: src/utils/task_stats_alloc/tests/loom.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+use loom::sync::atomic::{fence, AtomicUsize, Ordering};
+use loom::sync::Arc;
+use loom::thread;
+use task_stats_alloc::{allocation_stat, BYTES_ALLOCATED};
+use tokio::runtime::Handle;
+
+#[repr(transparent)]
+#[derive(Clone, Debug)]
+pub struct TaskLocalBytesAllocated(Option<NonNull<AtomicUsize>>);
```
Comment (wangrunji0408) [inline src/utils/task_stats_alloc/tests/loom.rs:19]: Is it possible to directly use the implementation from lib.rs and add `#[cfg]` to switch dependencies?

```rust
#[cfg(loom)]
use loom::sync::Arc;
#[cfg(not(loom))]
use std::sync::Arc;
```

File: src/utils/task_stats_alloc/tests/loom.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        let bytes_ref = self.0.as_ref().expect("bytes is invalid");
+        let bytes_ref = unsafe { bytes_ref.as_ref() };
+        bytes_ref.load(Ordering::Relaxed)
+    }
+}
+
+#[test]
+fn test_to_avoid_double_drop() {
+    loom::model(|| {
+        let bytes_num = 3;
```
Comment (liurenjie1024) [inline src/utils/task_stats_alloc/tests/loom.rs:76]: Can we try to increase the thread number?

**Review by wangrunji0408 — APPROVED** [review]
LGTM! 🥰

File: src/utils/task_stats_alloc/tests/loom.rs (hunk 2 above)
Comment (BowenXiao1999) [reply src/utils/task_stats_alloc/tests/loom.rs:19]: Yes it's possible. 

I plan to do this when we re-use `TaskLocalBytesAllocated` in the lib. Here this one is just a copy 😆

**Review by liurenjie1024 — APPROVED** [review]
LGTM

**Comment by liurenjie1024** [comment]
I think after this we should add daily tests on arm. cc @huangjw806 @sumittal

File: src/utils/task_stats_alloc/tests/loom.rs (hunk 3 above)
Comment (BowenXiao1999) [reply src/utils/task_stats_alloc/tests/loom.rs:76]: No. When I change from 3 to 4, the process just get aborted, can not get useful info.

There is a MAX_THREAD constants in loom crate. Maybe we can only max spawn 3 threads + original one = 4
```rust
/// Maximum number of threads that can be included in a model.
pub const MAX_THREADS: usize = 4;
```

Comment (wangrunji0408) [reply src/utils/task_stats_alloc/tests/loom.rs:76]: This limitation is much lower than what I expected it can do. 🥵

**Comment by huangjw806** [comment]
> I think after this we should add daily tests on arm. 

Sure, we can add it to our future plans.

**Pull request closed without merging**